### PR TITLE
[FSM] Implement graph traits for `fsm.machine`

### PR DESCRIPTION
**Pull request by mortbopet**
Initial implementation of LLVM graph traits for the `fsm.machine` operation. This will probably be a fairly critical piece of code for building further transformations, visualizations, ... upon.

Example `dot` output of the included `print-graph.mlir` test:
out

### CODE REVIEW

**Comment by mortbopet** [comment]
Clang tidy complains about things related to the (fixed) expected names for graph traits, so I think we can ignore those.

**Comment by teqdruid** [comment]
A few high level questions:
- Have you seen the MLIR `view-op-graph` pass?
- Would it make sense to expose this as a generic API?
- Would it make sense to have a 'graphviz' dialect?

**Comment by darthscsi** [comment]
> Clang tidy complains about things related to the (fixed) expected names for graph traits, so I think we can ignore those.

yes, several ancient llvm things are not clang-tidy clean.

**Review by darthscsi — APPROVED** [review]
Looks good to me.

File: test/Dialect/FSM/print-graph.mlir [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// RUN: circt-opt --fsm-print-graph %s -o %t 2>&1 | FileCheck %s
+
+// CHECK: [[IDLE:.*]] [shape=record,label="{IDLE|fsm.output %true : i1}"];
+// CHECK: [[IDLE]] -> [[BUSY:.*]][label="fsm.return %arg0"];
+// CHECK: [[BUSY]] [shape=record,label="{BUSY|fsm.output %false : i1}"];
+// CHECK: [[BUSY]] -> [[BUSY]][label="%0 = arith.cmpi ne, %cnt, %c0_i16 : i16\nfsm.return %0"];
+// FIX ME: FileCheck doesn't want to play nice with the following line (intended check: "[[BUSY]] -> [[IDLE]][label="%0 = arith.cmpi eq, %cnt, %c0_i16 : i16\nfsm.return %0"];")
```
Comment (mikeurbach) [inline test/Dialect/FSM/print-graph.mlir:7]: Can you use CHECK{LITERAL} to help with this?

https://llvm.org/docs/CommandGuide/FileCheck.html#directive-modifiers

**Review by mikeurbach — APPROVED** [review]
Makes sense, and I think packaging this up for use as an analysis is a good approach to start.

**Comment by mikeurbach** [comment]
> Would it make sense to have a 'graphviz' dialect?

I've thought about that a bit.. that might be neat. Another idea I had was to come up with some OpInterface that could enable dot emitters. Sort of like the LLVM graph traits but more MLIR flavored. The idea would be to do something similar to the printed form of the IR, where a dot emitter could handle MLIR structures generically, and if you implement the interface, various customizations could be enabled. Instead of each dialect writing a conversion to a dot dialect, each dialect would implement the various methods of the interface.

**Comment by mortbopet** [comment]
> Have you seen the MLIR view-op-graph pass?

I was not aware! However, taking it for a spin on this example generates something a bit too far from what i'd like to see:
image

> Would it make sense to have a 'graphviz' dialect?

I think I'd be more in favor of a set of `OpInterface`s, as @mikeurbach mentions. For a dialect - that is, an IR - to be warranted, I'd see it as a requirement that it is needed to do some form of transformations. I don't know what transformations I would do on a GraphViz IR since I'd only use it for dumping the graph itself. No matter what, _something else_ other than graph traits that is more MLIR-aware seems like a good idea to me - never been a big fan of graph traits!

File: test/Dialect/FSM/print-graph.mlir (hunk 1 above)
Comment (mortbopet) [reply test/Dialect/FSM/print-graph.mlir:7]: I don't think that would apply - the `BUSY` and `IDLE` variables capture names generated from the runtime addresses of the underlying nodes (default DOT graph traits behaviour) and thus change on every run with the same input.

**Pull request merged**